docs/blog/posts/write-code-to-refactor.md

```diff
@@ -0,0 +1,402 @@
+```
+
+This class is **very prone to errors** because we're relying on a single string to determine the logic. Imagine how much more complex we'd need to make the control flow when we add more options and complex configurations. 
+
+In order to prevent these complex errors from happening, we can use a `@classmethod` and a `enum` in order to prevent this from happening. A  `classmethod` allows us to instantiate a class with the right configuration so that we don't run into these complex edge cases. Furthermore, by pre-defining the values that a user can pass in, we can also support more complex initialisation logic.
+
+```python
+import enum
+
+class Provider(enum.Enum):
+    HUGGINGFACE = "HuggingFace"
+    OPENAI = "OpenAI"
+    COHERE = "Cohere"
+
+class EmbeddingModel:
+    def __init__(self,model_name:str,model_type=Provider):
+        self.model_name=model_name
+        self.model_type =model_type
+	
+    @classmethod
+    def from_openai(cls, model_name: str):
+        return cls(model_name, model_type=Provider.OPENAI)
+    
+    @classmethod
+    def from_hf(cls, model_name: str):
+        return cls(
+            model_name,
+            provider=Provider.HUGGINGFACE,
+        )
+```
+
+Small things like class methods and enums go a long way in ensuring that code that we write is safe and works as intended.
+
+### Write Simple Functions
+
+The functions I write have ideally three characteristics
+
+- They do one thing and their name reflects it
+- They are used at the same level of abstraction
+- They are parameterized with nice type hints 
+
+#### Do one thing well 
+
+A lot of times, it's very tempting to write a function that does a lot in one step. In the example above, we can see that we're trying to both format a specific chunk of data and then save it into the database.
+
+```python
+def format_data(data):
+  headers = extract_headers(data)
+	data = format_dataframe(data,headers)
+	upload_to_database(data)
+  return data
+
+def process():
+	data = get_data()
+	data = format_data(data)
+```
+
+There are two main problems here
+
+1. We're introducing a strange side effect that `format_data` doesn't reflect
+
+2. We're mutating state, which makes it difficult to debug issues when our logic gets more complicated
+
+A better way to write this logic would include splitting up the saving of data and the formatting of data into two separate functions. That way our functions closely reflect the actual actions that are being performed. We'd also try to minimize the mutation of state so that it's easier to trace errors when they happen
+
+```python
+def format_data(data):
+  headers = extract_headers(data)
+	return format_dataframe(data,headers)
+
+def save_data(data):
+  upload_to_database(data)
+
+def process():
+	data = get_data()
+	formatted_data = format_data(data)
+  save_to_database(data)
+```
+
+#### Be consistent with abstraction
```

Comment:
I don't think this is a good example, but we can really talk about it in terms of keeping the functions clear and simple.